src/lib.rs

```diff
@@ -521,17 +487,86 @@ pub enum LocParseError {
     SecondTabulatorMissing,
 }
 
-/// A brief in-app documentation.
-///
-/// This function will write a brief usage information, including a short
-/// introduction to the meaning of the configured `options`, to `stdout`.
-pub fn usage(options: &getopts::Options) {
-    let description = "Parses the Git history.";
-    let name = "commit-analyzer";
-    let synopsis = "[OPTIONS]";
-
-    println!(
-        "{name}.\n{description}\n\n{}",
-        options.usage(&format!("Usage: {name} {synopsis}"))
-    );
+/// Parses the Git history.
+#[derive(Parser, Debug)]
+#[clap(version, about, long_about = None)]
+pub struct Args {
+    #[clap(subcommand)]
+    input_method: InputMethod,
+
+    /// Always show the entire output.
+    #[clap(short, long)]
+    verbose: bool,
+
+    /// Filter for certain author names. ORs if specified multiple times.
+    #[clap(short, long)]
+    author_contains: Vec<String>,
+
+    /// Filter for certain author names. ORs if specified multiple times.
+    #[clap(long)]
+    author_equals: Vec<String>,
+
+    /// Filter for certain author emails. ORs if specified multiple times.
+    #[clap(short, long)]
+    email_contains: Vec<String>,
+    /// Filter for certain author emails. ORs if specified multiple times.
+    #[clap(long)]
+    email_equals: Vec<String>,
+
+    /// Filter for certain commit hashes. ORs if specified multiple times.
+    #[clap(short, long)]
+    commit_contains: Vec<String>,
+    /// Filter for certain commit hashes. ORs if specified multiple times.
+    #[clap(long)]
+    commit_equals: Vec<String>,
+
+    /// Filter loc for certain file extension (e.g. `--file-extension cpp`). ORs if specified multiple times.
+    #[clap(short, long)]
+    file_extension: Vec<String>,
+
+    /// Filter for certain commit messages. ORs if specified multiple times.
+    #[clap(short, long)]
+    message_contains: Vec<String>,
+    /// Filter for certain commit messages. ORs if specified multiple times.
+    #[clap(long)]
+    message_equals: Vec<String>,
+    /// Filter for certain commit messages. ORs if specified multiple times.
+    #[clap(short = 'l', long)]
+    message_starts_with: Vec<String>,
+
+    /// The time which may pass between two commits that still counts as working.
+    #[clap(short, long, default_value_t = 3)]
+    duration: u32,
+
+    /// An output file for the commits per day in CSV format.
+    #[clap(short, long)]
+    output: Option<PathBuf>,
+}
+
+impl Args {
+    /// Get a reference to the input method specified by the user.
```

Comment:
Yeah, I just took the auto generated documentation from rust analyzer and tried to improve it a bit. Will change.

Reply:
Thanks!  By the way another more general thing:  would you please move the struct before the `Author` definition and implementation?  That way, it is easier to take a look at the configured options *and* the structs are sorted by alphabet, again.

Reply:
Decided to move not only `Args` to the top, but also `InputMethod` (since this defines the subcommand, and so also the cli) and `Filter` since Filter is also very closely related to the `Args` supplied by the user. So I ordered them by relevance, instead of alphabetical.

Reply:
Please have a look at the following files:

* https://github.com/sorairolake/sysexits-rs/blob/develop/src/lib.rs
* https://github.com/sorairolake/sysexits-rs/blob/develop/src/exit_code.rs

These are the major source files of the `sysexits` crate we use as a dependency.

The maintainer organises their source code by using `lib.rs` for configuration and imports and outsources the other logic to dedicated source files in order to group related aspects.

I agree with you that the idea of grouping `Args` and `InputMethod` is a meaningful change since they are related.  But in contrast, violating against the alphabetical ordering just because those two data structures are related does not appear helpful to me.  At least in my opinion, this makes the overall navigation within the source file more complicated.

As an appropriate solution, I would like to suggest to resolve the arising issue of a harder navigation by applying at least one of the following strategies:

* Reflect the grouping in the documentation.  State in the file documentation how the data structures are ordered to outline your design decision.  The file documentation of the library crate root would thereby become something like a table of contents.
* Describe the relationship between the concerning data structures in their own documentation and outline the importance of them being situated next to each other in the source file.
* Outsource related data structures and logic to dedicated source files like the `sysexits` crate does.

Reply:
What are the advantages in your opinion of alphabetical ordering in this context?

Reply:
Its is easier to navigate through the code since you can rely on how the code is ordered.  Remember, not any editor offers you the possibility to jump to a certain scope to be chosen from a drop-down menu.

Reply:
The other way around would be to separate the logic from `src/lib.rs` to source files on their applying the intended grouping.  You can then use the `src/lib.rs` for configuration.  Example:  https://github.com/kevinmatthes/cargo-optimise/blob/main/src/lib.rs

Reply:
Since our `src/lib.rs` has grown quite big, I would like suggest to give the latter option a try.  Please consider to at least outsource the structs you grouped with your changes to a dedicated source file.  That way, their relationship becomes obvious and you can sort them by alphabet, again, in their source file.

Reply:
> One: That's too bad for the person using the editor, maybe they should switch.

Now, I am very really interested:  tell me, which command line editor for Linux has scope selection like Visual Studio Code-like IDEs offer when a corresponding Language Server is installed, and supports at least the following languages ***WITH syntax highlighting***:

* Lox
* Haskell
* Fortran
* Asymptote
* GNU Octave
* GLSL
* Godot Script
* LaTeX

I admit that Lox is a very pathological example but I think you get the idea:  there are also people who do not use GUI-oriented editors.  Regarding Lox, I found an editor, https://github.com/curlpipe/ox, which allows adding syntax highlighting rules.  Also https://github.com/zee-editor/zee has options to add parsers, but Haskell is still in development and there is no Fortran at all.

Which terminal editor do you use?  Yours seems to have scope selection and searching with `Ctrl`+`F`.

> Second: Every editor has Ctrl+F, which easily allows finding members in a file.

`mcedit` does not:  the search function is triggered with `F7`.  Also `vi` / `vim` and `nano` do not support exactly this key binding but have search functionality at all.

> In general I'm not convinced that alphabetical ordering adds anything to a source file, [...]

Luck you.  I saw that much ***really*** bad source code in my life that I am glad that I can at least rely on something like how the code is ordered if it does not even compile without being rejected / complained about by multiple compilers, parsers, linters and Language Servers.  And the documentation system -- if it has a documentation to be created at all.  Such code is extremely hard to read and almost not maintainable.  If cannot even rely on the correctness of the code such that it produces the expected output, there is no chance editing it.

In summary:  any source file should have several points you can rely on.  Its sorting is one of them.

Reply:
> I do not know why you need this restriction.. :sweat_smile: You are also allowed to use multiple editors for different languages :smile:

Restrictions are always depending on the perspective.  I consider an editor which depends on a window server, this is the desktop environment of your system, a restriction since you cannot use it in SSH sessions or when the corresponding service is either down or does not exist at all.  Go to the first available TTY of your Linux system and try to edit a file.  If you should use Windows, imagine how a system would feel that only has a `cmd` or PowerShell as the sole possibility of interaction -- no mouse, only your keyboard can be used to enter commands.  If you just have a TTY and no window server available, you cannot start your text editor.  You should make sure to be able to work with one if you need to, for instance in case of an error in the configuration files of your desktop environment or if the GPU drivers are missing.

Using multiple editors for multiple languages just because of their syntax highlighting is not an option all.  Imagine `vi` / `vim` enthusiasts were forced to use Emacs just because of its syntax highlighting or vice versa.  Either an editor is capable of highlighting the syntax of all the languages you need or it is not an option to use at all.  The idea of an editor is not only to support you entering the text and highlighting the text, if possible, but also to be the one-and-only tool for editing texts.  If you come from Emacs, are you sure that you want to learn all the `vi` / `vim` command modes just for ***ONE*** language you cannot plugin into Emacs?  I suppose you will not do so.

> Adding alphabetical ordering would add nothing to a file, that does not compile. It would even take away the original intention of the author and make the code less readable.

Well, either you sort the structs or I am going to do so in my next Pull Request.  I do not see any reason to *not* sort the structs by alphabet.

Reply:
> > I do not know why you need this restriction.. 😅 You are also allowed to use multiple editors for different languages 😄
> 
> Restrictions are always depending on the perspective. I consider an editor which depends on a window server, this is the desktop environment of your system, a restriction since you cannot use it in SSH sessions or when the corresponding service is either down or does not exist at all.

Then again, you could just use multiple editors depending on the occasion.. 😅 

> Go to the first available TTY of your Linux system and try to edit a file. If you should use Windows, imagine how a system would feel that only has a `cmd` or PowerShell as the sole possibility of interaction -- no mouse, only your keyboard can be used to enter commands. If you just have a TTY and no window server available, you cannot start your text editor. You should make sure to be able to work with one if you need to, for instance in case of an error in the configuration files of your desktop environment or if the GPU drivers are missing.

In most cases something similar to nano does completely suffice.

And if GPU drivers are missing, I recommend to first get those running, instead of programming in all those languages you listed 😄 

> Using multiple editors for multiple languages just because of their syntax highlighting is not an option all. Imagine `vi` / `vim` enthusiasts were forced to use Emacs just because of its syntax highlighting or vice versa. Either an editor is capable of highlighting the syntax of all the languages you need or it is not an option to use at all. The idea of an editor is not only to support you entering the text and highlighting the text, if possible, but also to be the one-and-only tool for editing texts. If you come from Emacs, are you sure that you want to learn all the `vi` / `vim` command modes just for _**ONE**_ language you cannot plugin into Emacs? I suppose you will not do so.

Well, if you have to be religious about your editor? It seems just to be a bad reason to narrow your horizon. In my experience everything, that claims to be a Swiss Army Knife without actually being one, is worse then the things actually designed to do the task at hand.
 
> > Adding alphabetical ordering would add nothing to a file, that does not compile. It would even take away the original intention of the author and make the code less readable.
> 
> Well, either you sort the structs or I am going to do so in my next Pull Request. I do not see any reason to _not_ sort the structs by alphabet.

Like I already explained, since `Args` and `InputMethod` are both interface to `clap` (and so is `Filter` in the way, that it is completely specified by the user), they are relevant to each other and should be close to each other. That is my reason, but I already explained it once. 😄

Reply:
In order to return to the original purpose of this discussion:  let us just agree on that the sorting of symbols in a source file is a matter of taste to be discussed more in detail and / or applied in another Pull Request or issue.